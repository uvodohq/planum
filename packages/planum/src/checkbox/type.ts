import type {
  AriaCheckboxGroupItemProps,
  AriaCheckboxGroupProps,
  AriaCheckboxProps,
  ToggleProps,
} from '@react-types/checkbox'
import type { ReactNode } from 'react'

import type { CSS } from '../theme'
import type { StyledCheckboxVariants } from './checkbox.styles'

export interface CustomProps {
  css?: CSS
  label?: string
  children?: string
}

export type CheckboxProps = CustomProps &
  StyledCheckboxVariants &
  ToggleProps &
  AriaCheckboxProps

export type CheckboxGroupItemProps = CheckboxProps & AriaCheckboxGroupItemProps

export interface CheckboxBaseProps {
  isDisabled?: boolean
  isIndeterminate?: boolean
  isSelected: boolean
  isFocusVisible?: boolean
  label?: any
  pressProps?: any
  inputProps?: any
  focusProps?: any
  children?: any
  inputRef?: any
}

export type CheckboxGroupProps = AriaCheckboxGroupProps & {
  children: ReactNode
  css?: CSS
}
