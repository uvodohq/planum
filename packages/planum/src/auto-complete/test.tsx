import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
