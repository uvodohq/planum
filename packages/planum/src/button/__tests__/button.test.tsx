import * as React from 'react'
import { vi } from 'vitest'

import { render, screen, userEvent } from '../../test'
import type { ButtonProps } from '../'
import { Button } from '../'

function renderButton(props?: ButtonProps) {
  return render(<Button {...props}>Button</Button>)
}

function getButton() {
  return screen.getByRole('button')
}

// useProgressBar generates a random id for the progress bar, breaks snapshots. so we mock it
vi.mock('@react-aria/progress', async () => {
  return {
    useProgressBar: vi.fn(() => ({
      progressBarProps: {
        role: 'progressbar',
      },
    })),
  }
})

describe('<Button />', async () => {
  it('should render the button', () => {
    renderButton()
    const button = getButton()

    expect(button).toBeInTheDocument()
    expect(button).toHaveTextContent('Button')

    expect(button).toMatchSnapshot()
  })

  it("should navigate '/test' when given as props and clicked", () => {
    renderButton({ as: 'a', href: '/test' })
    const button = screen.getByRole('link')

    expect(button).toBeInTheDocument()
    expect(button).toHaveTextContent('Button')

    userEvent.click(button)

    expect(button).toHaveAttribute('href', '/test')
  })

  it('should call onClick handler', async () => {
    const onClickMock = vi.fn()
    renderButton({
      onClick: onClickMock,
    })
    const button = getButton()

    expect(onClickMock).not.toHaveBeenCalled()
    await userEvent.click(button)
    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  it('should not call onClick callback if disabled', async () => {
    const onClickMock = vi.fn()
    renderButton({
      onClick: onClickMock,
      isDisabled: true,
    })
    const button = getButton()

    expect(button).toBeDisabled()
    await userEvent.click(button)
    expect(onClickMock).not.toHaveBeenCalled()

    expect(button).toMatchSnapshot()
  })

  it.only('should show loading and not accept onClick', async () => {
    const onClickMock = vi.fn()
    renderButton({
      onClick: onClickMock,
      isLoading: true,
    })
    const button = getButton()
    const loading = screen.getByRole('progressbar')

    expect(button).toBeDisabled()
    expect(loading).toBeInTheDocument()

    await userEvent.click(button)
    expect(onClickMock).not.toHaveBeenCalled()

    expect(button).toMatchSnapshot()
  })

  it('should call click callback when focused with tab and enter key press', async () => {
    const onClickMock = vi.fn()
    renderButton({
      onClick: onClickMock,
    })
    const button = getButton()

    expect(onClickMock).not.toHaveBeenCalled()
    await userEvent.tab()

    expect(button).toHaveFocus()

    await userEvent.keyboard('{enter}')
    expect(onClickMock).toHaveBeenCalledTimes(1)

    expect(button).toMatchSnapshot()
  })

  it('should call click callback when focused with tab and spacebar key press', async () => {
    const onClickMock = vi.fn()
    renderButton({
      onClick: onClickMock,
    })
    const button = getButton()

    expect(onClickMock).not.toHaveBeenCalled()
    await userEvent.tab()

    expect(button).toHaveFocus()

    await userEvent.keyboard('{ }')
    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  it('should focus when tab key press', async () => {
    renderButton()
    const button = getButton()

    expect(button).not.toHaveFocus()
    await userEvent.tab()
    expect(button).toHaveFocus()
  })
})
