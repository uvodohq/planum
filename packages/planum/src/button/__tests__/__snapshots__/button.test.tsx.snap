// Vitest Snapshot v1

exports[`<Button /> > should show loading and not accept onClick 1`] = `
<button
  aria-busy="true"
  aria-label="Loading"
  class="Planum-c-eDYXqN Planum-c-eDYXqN-KUnxm-size-md Planum-c-eDYXqN-eEPmHB-variant-primary Planum-c-eDYXqN-jUrwBA-isLoading-true"
  disabled=""
  type="button"
>
  <span
    class="Planum-c-kxAGWb Planum-c-kxAGWb-icKbjjX-css"
  >
    <span
      class="Planum-c-dbzzBd"
    >
      Button
    </span>
  </span>
  <span
    class="Planum-c-gXYcWj"
  >
    <span
      class="Planum-c-ihyEMT Planum-c-ihyEMT-fEcTqV-speed-fast Planum-c-ihyEMT-hRvkRI-size-default Planum-c-ihyEMT-eGYJCk-theme-dark"
      role="progressbar"
      style="--planum-loader-percentage: 100;"
    >
      <svg
        class="Planum-c-dIfLHp"
        viewBox="0 0 50 50"
      >
        <circle
          class="Planum-c-jlBdpC"
          cx="25"
          cy="25"
          fill="none"
          r="22"
          stroke-width="4"
        />
        <circle
          class="Planum-c-PJLV"
          cx="25"
          cy="25"
          fill="none"
          r="22"
          stroke-width="4"
        />
      </svg>
    </span>
  </span>
</button>
`;
